docs/developer/designs/node-sets/README.md

```diff
@@ -0,0 +1,159 @@
+- Plugin - A piece of code extending KAI scheduler native behavior by implementing custom logic
+- Hooks - Extension points for plugin developers, providing multiple points for affecting scheduling decisions
+- Session - A runtime object representing the scheduling cycle, exposing functions to activate registered plugins
+- Topology Scheduling Plugin - A plugin implementing topology scheduling logic
+
+## High Level
+
+### NodeSet
+Introduce new type of `NodeSet`
+
+```go
+type NodeSet []*NodeInfo
+```
+
+This type is defined as a zero or more nodes (Single NodeSet equals to multiple nodes grouped together).
+
+### SubsetNodes hook
+Introduce the new hook that for a given PodGroup returns a list of NodeSets
+```go
+type SubsetNodesFn func(PodGroup) []NodeSet
+```
+
+Such function accept PodGroup as argument and provides multiple NodeSets to attempt scheduling over.
+By implementing such function you can register your code as SubsetNodes plugin.
+
+A node can be in a zero or more NodeSets.
+
+## Defined behavior
+
+### Registration 
+Like any other hook, the developer require to register the plugin prior to compilation.
+```go
+framework.RegisterPluginBuilder("myPlugin", myPlugin.New)
+```
+
+In addition the user should add the plugin name into the scheduler configuration.
+
+### Session function
+The session object expose API for activating the plugins that implement the hook.
+
+```go
+func (ssn *Session) SubsetNodes(PodGroup) []NodeSet
+```
+
+Such function is accessible during cycle, as long as the session exists, and exposes the logic of all the implementing plugins.
+
+In case of multiple plugins that implement that hook, the expected behavior is multiple layer of SubSetting.
+For example if you have 2 plugins `P` and `R`. Dividing the cluster nodes happens in 2 steps:
+
+First plugin, `P`, divide all cluster nodes to subsets:
+```
+AllNodes -> P1, P2, P3
+```
+Then plugin R, divide each subset:
+```
+P1 -> P1R1, P1R2
+P2 -> P2R1, P2R2, P2R3
+P3 -> P3R1, P3R2
+```
+So the final results of subsets is:
+```
+P1R1, P1R2, P2R1, P2R2, P2R3, P3R1, P1R2
+```
+
+
+### During scheduling
+
+During scheduling when a PodGroup is chosen to be scheduled, we call that hook before attempting to allocate any task, dividing the cluster nodes to NodeSets.
+
+As opposed to traditional behavior, we will now attempt to schedule the PodGroup over each NodeSet, until we find one over its scheduled, and stop the iteration.
+
+```go
+for podGroup := range PodGroups {
+	nodeSets := ssn.SubsetNodes(podGroup)
+	for nodeSet := range nodeSets {
+		if attemptToAllocatePodGroup(podGroup, nodeSet) {
+			break
+        }   
+    }
+}
+```
+It means that for each PodGroup we perform the division, and then try each NodeSet, until we find a matching one.
+
+
+## Low level design
+
+```go
+package NodeInfo
+
+type NodeSet []*NodeInfo
+```
+
+```go
+package api
+
+type SubsetNodesFn func(job *podgroup_info.PodGroupInfo, isRealAllocation bool, nodeSet node_info.NodeSet) error
```

Comment:
Instead of passing `isRealAllocation` directly, maybe create a struct to deliver other relevant parameters? Do we expect any such parameters with the podgroup subsets?

Reply:
Actually the only reason we need `isRealAllocation` parameter is in order to count how many tasks needs to be allocated (Pending only / virtually evicted pods also).

Another option, tell me what you think, is instead of passing job + isRealAllocation, is passing job + List of tasks.
And someone from outside will decide which tasks to pass based weather is real allocation or simulation

Reply:
I am also in favor of passing a job with a list of tasks